Fixes Prepare > Column: Numeric > Row Statistics dialogue  Row statistics dialogue

EstherNjeriLiberatta opened the pull request:
Fixes #7240 
This is still a work in progress.

EstherNjeriLiberatta commented:
@rdstern @lilyclements I have done the first part of the changes you requested with help from @N-thony i.e.
a.) Two radio buttons at the top. Label first, namely `Summaries: `. Then` Single` and` Multiple`
b) The Single option is what we have now, namely each row of data gives a single summary. 
c) In Single swap the Count and Standard Deviation. Also change Number of Missing to Number Missing
d) Get rid of the User Defined button, but have a 9th radio button labelled User Defined.
e) If User defined is chosen, then open a drop down with a text field into which you can type a command. The drop down can include (some of) the 8 existing ones, plus a few more to be decided.
f) Now the code generated. First, the existing code has a very long first line. Could you separate getting the data, from using it. 

Kindly review the work for any changes. Thank you.

rdstern reviewed (changes requested):
@EstherNjeriLiberatta this looks good now.  I wonder about merging this version, and then attacking the other parts I put in the initial issue, and have now added to.  
So, could you just disable the Multiple button for now.  Then, in parallel, please could @shadrackkibet or @lloyddewit check the code, so it can be merged.  
I note (that's why I didn't see it) that the missing option is not visible, when you choose user defined.  That's fine now, but in the improvements, after merging, it should be available then - and included in the commands that will be generated.  (Not now, because the current list will change.)

shadrackkibet commented on instat/dlgRowSummary.vb:
```diff
@@ -120,5 +137,43 @@ Public Class dlgRowSummary
+        clsGetColumnsFunction = ucrReceiverForRowSummaries.GetVariables(True).Clone
+        clsGetColumnsFunction.SetAssignTo("columns")
+        clsApplyFunction.AddParameter("X", clsRFunctionParameter:=clsGetColumnsFunction, iPosition:=0)
+        ucrBase.clsRsyntax.AddToBeforeCodes(clsGetColumnsFunction)
```
Perhaps I didn't understand the reason why this is done this way. Do you really need `clsGetColumnsFunction` function?  Isn't it just 

```
        ucrReceiverForRowSummaries.bForceAsDataFrame = True
        ucrReceiverForRowSummaries.SetParameterIsRFunction()
```

EstherNjeriLiberatta replied:
It was helpful when I was trying to separate the first line of code from the main function. 
 `ucrReceiverForRowSummaries.SetParameterIsRFunction()` didn't work as expected.

shadrackkibet commented on instat/dlgRowSummary.vb:
```diff
@@ -120,5 +137,43 @@ Public Class dlgRowSummary
         Else
             clsApplyFunction.RemoveParameterByName("na.rm")
         End If
+        If rdoMore.Checked Then
+            clsOtherDummyFunction.AddParameter("checked", "user_defined", iPosition:=0)
+            clsApplyFunction.AddParameter("FUN", ucrInputUserDefined.GetText, iPosition:=2)
+        ElseIf rdoMean.Checked Then
+            clsOtherDummyFunction.AddParameter("checked", "mean", iPosition:=0)
+            clsApplyFunction.AddParameter("FUN", "mean", iPosition:=2)
+        ElseIf rdoMaximum.Checked Then
+            clsOtherDummyFunction.AddParameter("checked", "max", iPosition:=0)
+            clsApplyFunction.AddParameter("FUN", "max", iPosition:=2)
+        ElseIf rdoMinimum.Checked Then
+            clsOtherDummyFunction.AddParameter("checked", "min", iPosition:=0)
+            clsApplyFunction.AddParameter("FUN", "min", iPosition:=2)
+        ElseIf rdoStandardDeviation.Checked Then
+            clsOtherDummyFunction.AddParameter("checked", "sd", iPosition:=0)
+            clsApplyFunction.AddParameter("FUN", "sd", iPosition:=2)
+        ElseIf rdoSum.Checked Then
+            clsOtherDummyFunction.AddParameter("checked", "sum", iPosition:=0)
+            clsApplyFunction.AddParameter("FUN", "sum", iPosition:=2)
+        ElseIf rdoMedian.Checked Then
+            clsOtherDummyFunction.AddParameter("checked", "median", iPosition:=0)
+            clsApplyFunction.AddParameter("FUN", "median", iPosition:=2)
+        ElseIf rdoCount.Checked Then
+            clsOtherDummyFunction.AddParameter("checked", "count", iPosition:=0)
```
- I also did some quick testing and got the below error when I selected `More` and in the drop-down selected `number missing`.
- This combo box should be non-editable.
- I also see the same summaries as the radio buttons repeated inside the combo box.

![image](https://user-images.githubusercontent.com/26301973/156141171-1722e274-25d8-4db1-9187-e67feac157ad.png)

shadrackkibet commented on instat/dlgRowSummary.vb:
```diff
@@ -35,34 +39,38 @@ Public Class dlgRowSummary
         ucrReceiverForRowSummaries.SetMeAsReceiver()
         ucrReceiverForRowSummaries.strSelectorHeading = "Numerics"
         ucrReceiverForRowSummaries.SetIncludedDataTypes({"numeric"})
-        ucrReceiverForRowSummaries.bUseFilteredData = False
```
Could you explain why was this removed?

EstherNjeriLiberatta replied:
@shadrackkibet I had commented on it before to see whether there was a difference in the code. I might have removed it when cleaning the code. I'll add it back, thank you.

rdstern commented:
@shadrackkibet said:
I also did some quick testing and got the below error when I selected More and in the drop-down selected number missing.
This combo box should be non-editable.
I also see the same summaries as the radio buttons repeated inside the combo box.

a) I asked (temporarily) for some of the same summaries.  But that will change in the next round, as specified in the issue (now).
b) Could delete the number missing, for now, if it gives an error.
c) The new items in the combo box will need the control to be editable.  So prefer that option to be kept.  (One example will be quantile,prob = 0.5, where the user could change that level to another value.

EstherNjeriLiberatta commented:
Hello, @shadrackkibet I've fixed the changes you requested. I've also added an omit missing values control with no code yet. I will continue adding more features to the dialogue once it's merged. Kindly review. Thank you.

rdstern commented:
@EstherNjeriLiberatta pleased you are back to this dialogue.  I was puzzled by your comment on the addition of the Omit Missing Values , because that is already there for some of the options, and seems to be working.  It isn't visible on the More option, but I presume could be made so?

EstherNjeriLiberatta commented:
@rdstern Yes it can, I will add it to the more option.

rdstern commented:
@EstherNjeriLiberatta has this version taken account of the rowwise suggestion from @lilyclements.  If not, then I would suggest you include that in the current pull request.  Then perhaps add at least some of the items suggested for the More before merging?  Good to discuss prehaps with @shadrackkibet if you need to.  Are you in Maseno, where he is now?

EstherNjeriLiberatta commented:
@rdstern Okay. Yes, he is here, I have discussed this with him. Thank you for the feedback.

rdstern commented:
@EstherNjeriLiberatta should others have taken over this work?  There has been silence for a long time.  Are you still working?

shadrackkibet commented:
@Vitalis95 @anastasia-mbithe @mmulahvictor I suggest one of you take over this task.

shadrackkibet commented:
@africanmathsinitiative/developers Just a follow-up on this. Any volunteer to take over?

EstherNjeriLiberatta commented:
@rdstern My apologies for the late reply. Yes, thank you for your patience. I'm working on this. @shadrackkibet I'll be working on this.

shadrackkibet commented:
@EstherNjeriLiberatta we are reassigning all your tasks at the moment. I would suggest you don't do anything on this PR. 

The way forward is - Get in touch with Zach to clarify your work situation and respond to emails.

EstherNjeriLiberatta commented:
@shadrackkibet Okay, thank you.

rdstern commented:
@EstherNjeriLiberatta there is still a fair bit to do on this dialogue, which I am very keen to include in the forthcoming update.  This includes the multiple option.  If you feel this may be too much work for you, then feel free to share with @N-thony.

EstherNjeriLiberatta commented:
Okay, @rdstern  I will share it with him to work faster, thank you.

EstherNjeriLiberatta commented:
@shadrackkibet I'm working on this with @N-thony.

EstherNjeriLiberatta commented:
@rdstern We've been having a challenge installing the matrixStats package on my laptop. @N-thony requesting your help to test the changes I've made. We are trying to fix the problem that's why it has been delayed but it should work okay on your end, just a few more adjustments on the dialogue. Thank you.

EstherNjeriLiberatta commented:
@rdstern This is ready for review. 
For testing I used the experimental_suvey dataset. You now have `rowwise()` instead of `apply()` for all functions in the single option as suggested by @lilyclements. New columns are no longer being added but instead you are updating the dataframe.
The Multiple option is using the `matrixStats` package. It has the three radio buttons: `Ranks`, `Ranges` and `Quantiles` all using `rowRanks()`, `row Ranges()`, and `rowQuantiles()` repsectively. You have additional combo boxes that is, `Ties` and `Type` for radio buttons `Ranks` and `Quantiles` respectively.

rdstern commented:
@anastasia-mbithe this is looking really good.  There was a lot of work here.  Great job so far.  Just a bit of tidying needed!
a) Delete the mfv function, which gives an error if there is more than 1 mode in the data.
b) 2 of the functions allow the user to type.  They are the trimmed mean and the quantile.  But when I type anything it doesn't work, it gives a null function in the code.  For example changing quantile 0.5 to 0.4 or trimmed mean with 0.1 instead of 0.2.
c) Now a bigger problem, but easy to fix I hope?  The extra variable should add a variable to the existing dataframe.  In each case now it now does this, but then in a new data frame.
d) Now the multiple options.  
e) The ranks works - that's great!  In the dialogue, move the Ties control over a bit and make it wider.  Could you code it so the resulting variable names start with the original name, and add _r, say, for rank?  For the ranks there are as many new variables as there were variables.
f) For the range could the initial names be r_min and r_max?  (Of course they will be r_min1 and r_max1 is they already exist.
g) For quantiles the first pull-down could be made quite a lot wider, and the second could be much narrower.  Could the names perhaps be q0.2_, etc.  
h) I tried my own quantiles and they seem to work - great!

This is very good progress!  Well done indeed.

EstherNjeriLiberatta commented:
> @anastasia-mbithe this is looking really good. There was a lot of work here. Great job so far. Just a bit of tidying needed! a) Delete the mfv function, which gives an error if there is more than 1 mode in the data. b) 2 of the functions allow the user to type. They are the trimmed mean and the quantile. But when I type anything it doesn't work, it gives a null function in the code. For example changing quantile 0.5 to 0.4 or trimmed mean with 0.1 instead of 0.2. c) Now a bigger problem, but easy to fix I hope? The extra variable should add a variable to the existing dataframe. In each case now it now does this, but then in a new data frame. d) Now the multiple options. e) The ranks works - that's great! In the dialogue, move the Ties control over a bit and make it wider. Could you code it so the resulting variable names start with the original name, and add _r, say, for rank? For the ranks there are as many new variables as there were variables. f) For the range could the initial names be r_min and r_max? (Of course they will be r_min1 and r_max1 is they already exist. g) For quantiles the first pull-down could be made quite a lot wider, and the second could be much narrower. Could the names perhaps be q0.2_, etc. h) I tried my own quantiles and they seem to work - great!
> 
> This is very good progress! Well done indeed.

@rdstern Thanks for your review, I will make the changes as requested but I guess tagged @anastasia-mbithe in this PR by mistake.

EstherNjeriLiberatta commented:
@rdstern, I noticed the issue you have mentioned in part **b**. It is one of the challenges I experienced. I'm working on fixing it by finding a way in Visual Studio so that when you change it, it also changes the R code. I'm looking into it with help from @N-thony. Thanks again for your review.

rdstern commented:
@EstherNjeriLiberatta ooops, you are right, I mentioned ana by mistake.

rdstern commented:
@EstherNjeriLiberatta just to confirm that the bottom control currently is:

![image](https://user-images.githubusercontent.com/11226469/163923806-0652185b-0fb6-4276-a514-2c810180fa77.png)

Currently that's the name of the new data frame - which we don't need.  The new column name is currently not row_summary, but the name of the summary itself, which is great and just what we need.  So the change may be in the control, but keep the column summary name as you have it now.

EstherNjeriLiberatta commented:
@rdstern, @Ivanluv is suggesting that we have an `nud` control for the parameters in the mean and quantile functions rather than typing in the input. It will look like this:
![Uploading User defined mean.jpg…]()

EstherNjeriLiberatta commented:
@rdstern @lilyclements This is ready for review.
I have made the changes you requested that is;
**For the Single Option**:
a) Delete the `mfv` function, which gave an error if there is more than 1 mode in the data.
b) Fixed the two functions that allow the user to type in i.e. trim mean and quantile.
c) The extra variable is adding a variable to the existing data frame. 
**For the multiple option**.
a) Moved the Ties control over a bit and make it wider. 
b) Coded it so the resulting variable names start with the original name, and added r, say, for rank.
c) For the range, coded the initial names to be r_min and r_max.
d) For quantile, made the first pull-down a lot wider, and the second much narrower. 
e)Coded the names to be q0.2, etc.

lilyclements commented on instat/dlgRowSummary.vb:
```diff
@@ -35,68 +48,357 @@ Public Class dlgRowSummary
+        ucrInputProbability.SetParameter(New RParameter("p", 1, bNewIncludeArgumentName:=False))
+        ucrInputProbability.AddQuotesIfUnrecognised = False
+        ucrInputProbability.SetValidationTypeAsNumericList()
+        ucrInputProbability.SetDropDownStyleAsNonEditable()
```
I think this should be editable still, so this line can be removed. We just need to have it in the TestOkEnabled to check it isn't empty (when the multiple rdo is checked)

lilyclements commented on instat/dlgRowSummary.vb:
```diff
@@ -35,68 +48,357 @@ Public Class dlgRowSummary
+            End If
         Else
-            ucrBase.OKEnabled(False)
+            If Not ucrReceiverForRowSummaries.IsEmpty Then
```
Should include ucrInputProbability here. So can change this to:
`            If Not ucrReceiverForRowSummaries.IsEmpty AndAlso ucrInputProbability.GetText <> "" Then`

And also add `ucrInputProbability.ControlContentsChanged` later for TestOK sub.

lilyclements replied:
@EstherNjeriLiberatta have you added this in?

lilyclements reviewed (commented):
Looks good! Just two very minor comments

rdstern reviewed (changes requested):
Looks good.
a) TRimmed mean make maximum 0.5 with steps of 0.01
b) Quantile ok from 0 to 1, but steps of 0.01.
c) Multiple, label, change quantile to quantiles

lilyclements commented on instat/dlgRowSummary.vb:
```diff
@@ -35,68 +48,354 @@ Public Class dlgRowSummary
+            End If
         Else
-            ucrBase.OKEnabled(False)
+            If Not ucrReceiverForRowSummaries.IsEmpty AndAlso ucrInputProbability.GetText <> "" Then
```
Change `AndAlso ucrInputProbability.GetText <> "" Then` to

`AndAlso ((rdoRowQuantile.IsChecked AndAlso ucrInputProbability.GetText <> "") OrElse Not rdoRowQuantile.IsChecked) Then`

This would mean that OK is disabled when the ucrInput is empty only if the ucrInput is visible (that is, if the rdoRowQuantile is checked)

lilyclements commented on instat/dlgRowSummary.vb:
```diff
@@ -106,19 +405,220 @@ Public Class dlgRowSummary
+    End Sub
+
+    Private Sub Controls_ControlContentsChanged(ucrChangedControl As ucrCore) Handles ucrReceiverForRowSummaries.ControlContentsChanged, ucrPnlStatistics.ControlContentsChanged,
+        ucrSaveNewDataFrame.ControlContentsChanged, ucrInputProbability.ControlContentsChanged, ucrInputUserDefined.ControlContentsChanged, ucrPnlMultipleRowSummary.ControlContentsChanged, ucrPnlRowSummaries.ControlContentsChanged
```
Add in here `ucrPnlMultipleRowSummary.ControlContentsChanged` due to suggested changes in the TestOK

lilyclements reviewed (commented):
Looks good! One or two final comments

EstherNjeriLiberatta commented:
Fixed the Nud to have an increment of 0.01 for both mean and quantile in the single option.
Changed the maximum value for the trimmed mean to 0.5.
Fixed bug on test okay.
@rdstern This is ready for review.

EstherNjeriLiberatta commented:
Changed Quantile to Quantiles on the multiple option. 
@rdstern @africanmathsinitiative/developers this is ready for review

rdstern reviewed (approved):
Great.  Really good if this could be included in the forthcoming version

lloyddewit commented:
@lilyclements Thank you for your review.
You have 4 open comments above. Please could you check if they've been satisfactorily resolved, and if so, click on their 'Resolve conversation' buttons.
Thanks!

lilyclements commented:
@lloyddewit thanks. I am happy that the comments I made are resolved.

rdstern commented:
@lloyddewit given @lilyclements comments above is this ok to approve now?

lloyddewit commented on instat/dlgRowSummary.vb:
```diff
@@ -106,19 +414,220 @@ Public Class dlgRowSummary
+                    clsDummyRowFunction.AddParameter("check", "rowRange", iPosition:=0)
+                    clsColumnNamesFunction.AddParameter("x", clsRFunctionParameter:=clsRowRangesFunction, iPosition:=0, bIncludeArgumentName:=False)
+                    ucrBase.clsRsyntax.SetBaseROperator(clsAssignOperator)
+
```
```suggestion
```
